test/utils/testFetch.ts

```diff
@@ -0,0 +1,52 @@
+import { TaskEither } from 'fp-ts/lib/TaskEither';
+import { fromNullable } from 'fp-ts/lib/Option';
+import { ObservableQuery } from '../../src/Query';
+import { QueryResult } from '../../src/QueryResult';
+import { observeShallow } from '../../src/observe';
+import { toArray, take, timeout } from 'rxjs/operators';
+import { identity } from 'fp-ts/lib/function';
+
+type TestFetch<A, L, P> = (input: A) => TaskEither<L, P>;
+
+export function makeTestFetch<A = void, L = unknown, P = unknown>(
+  results: Array<TestFetch<A, L, P>>
+): jest.Mock<TaskEither<L, P>, [A]> & { assertExhausted: () => void } {
+  let index = 0;
+  const fn = jest.fn(input => {
+    const te = fromNullable(results[index])
+      .map(f => f(input))
+      .foldL(() => {
+        throw new Error('test fetch out of range');
+      }, identity);
+    index += 1;
+    return te;
+  });
+
+  (fn as any).assertExhausted = () =>
```

Comment:
I feel that more often than not we don't want to check that we called the fetch function as many times as the options we gave it (this is already covered in part by `throw new Error('test fetch out of range')`) but most of the times we want to check the exact number of times it was called

Reply:
this is bad naming, but the check performed in the mock impl. is actually `called # times === expected.length`, so as you are suggesting. We are passing expected invocations, we want to check that each all are actually used.

Reply:
what I meant is more that `expected.lenght` and `results.lenght` should be different, with the current implementation if you call the fetch function a number of times > than `results.lenght` we throw (but I feel that is not what we want)

Reply:
I suggest we take this one offline :)

Reply:
agreed :)